### PR TITLE
feat: implement LSTM and GRU operators for torchlib

### PR DESCRIPTION
**Pull request by ombrdr47**
Implement aten_lstm and aten_gru operators to enable torch.onnx.export for PyTorch LSTM and GRU layers. This addresses issue #2546.

Key features:
- Full support for multi-layer RNNs (num_layers > 1)
- Bidirectional support (forward and backward directions)
- Handles both biased and non-biased configurations
- batch_first parameter support with automatic transposition
- Dropout support between layers (nondeterministic seeded)
- Proper gate reordering for ONNX compatibility:
  * LSTM: PyTorch [i,f,g,o] -> ONNX [i,o,f,g]
  * GRU: PyTorch [r,z,n] -> ONNX [z,r,n]

Implementation details:
- Uses ONNX LSTM/GRU operators with proper parameter formatting
- Handles weight matrix transposition and reshaping
- Correctly concatenates biases using op.Concat
- Processes each layer independently with proper state management
- Returns outputs in PyTorch-compatible format

Closes: #2546

Also resolves:
- pytorch/pytorch#120626 (GRU export)
- pytorch/pytorch#123089 (LSTM export)  
- pytorch/pytorch#164834 (GRU dynamo export)

### CODE REVIEW

**Comment by ombrdr47** [comment]
@microsoft-github-policy-service agree

File: onnxscript/function_libs/torch_lib/ops/core.py [hunk 1]
```diff
@@ -4991,6 +4991,356 @@ def aten_lstm_mps_backward(
     raise NotImplementedError()
 
 
+@torch_op("aten::lstm", trace_only=True)
```
Comment (justinchuby) [inline onnxscript/function_libs/torch_lib/ops/core.py]:
```suggestion
@torch_op("aten::lstm.input", trace_only=True)
```

File: onnxscript/function_libs/torch_lib/ops/core.py [hunk 2]
```diff
@@ -4991,6 +4991,356 @@ def aten_lstm_mps_backward(
...
+    
+    # Handle batch_first for output
+    if batch_first:
+        # Convert from [seq, batch, features] to [batch, seq, features]
+        current_input = op.Transpose(current_input, perm=[1, 0, 2])
+    
+    return current_input, final_h, final_c
+
+
+@torch_op("aten::gru", trace_only=True)
```
Comment (justinchuby) [inline onnxscript/function_libs/torch_lib/ops/core.py]:
```suggestion
@torch_op("aten::gru.input", trace_only=True)
```

Comment (justinchuby) [inline onnxscript/function_libs/torch_lib/ops/core.py]: Could you put this function with the `aten_g...` functions so that they are sorted alphabetically?

Comment (ombrdr47) [reply onnxscript/function_libs/torch_lib/ops/core.py]: Done! Moved `aten_gru` to appear after `aten_ger`.

**Comment by ombrdr47** [comment]
> Feel free to add commit the tests and I can take a look

Thanks! I've added all 6 tests and pushed them. The LSTM tests all pass, but the GRU tests are failing. The outputs don't match PyTorch within tolerance even though the gate reordering looks correct.
All changes are commited. Thank you!

File: tests/function_libs/torch_lib/e2e_ops_tests.py [hunk 3]
```diff
@@ -302,6 +302,90 @@ def forward(self, x):
...
+            def __init__(self):
+                super().__init__()
+                self.gru = torch.nn.GRU(input_size=10, hidden_size=20, num_layers=3, batch_first=True)
+
+            def forward(self, x):
+                return self.gru(x)
+
+        model = GRUModel()
+        x = torch.randn(5, 3, 10)  # (batch, seq, input_size)
+        onnx_program = torch.onnx.export(model, (x,), dynamo=True, verbose=False)
```
Comment (titaiwangms) [inline tests/function_libs/torch_lib/e2e_ops_tests.py:406]: Could you also add some cases to test the dynamism? I think batch_size and sequence length should at least be dynamic?

For example (I might be wrong on axis names), 
```python
onnx_program = torch.onnx.export(
    model, 
    (x,),
    dynamic_shapes=({0: "batch", 1: "seq_len"}) 
    dynamo=True,
    verbose=False
)
```

**Comment by justinchuby** [comment]
I think it's ok to skip dynamic shapes testing for now if the error comes from torch.export, thanks!

**Comment by titaiwangms** [comment]
> I think it's ok to skip dynamic shapes testing for now if the error comes from torch.export, thanks!

Have not seen this error message for a while. Is the CI using old torch? (before oblivious backed size?)

**Comment by justinchuby** [comment]
> > I think it's ok to skip dynamic shapes testing for now if the error comes from torch.export, thanks!
> 
> Have not seen this error message for a while. Is the CI using old torch? (before oblivious backed size?)

Right we do test with 2.7
